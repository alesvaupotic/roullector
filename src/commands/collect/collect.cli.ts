import {
  Command,
  Option,
} from 'commander';

import { collect } from '$commands/collect/collect';
import {
  defaultCliCollectOptions,
  keyTransformCliToFunc,
} from '$commands/collect/collect.constants';
import type {
  CollectOptions,
  CliCollectOptions,
} from '$commands/collect/collect.types';

export function collectCli() {
  return new Command('collect')
    .description('collect route data')
    .option(
      '-i, --inDir <path>',
      'directory path to collect',
      defaultCliCollectOptions.inDir
    )
    .option(
      '-e, --extensions <list>',
      'comma-separated list of file extensions to collect',
      defaultCliCollectOptions.extensions,
    )
    .option(
      '-x, --ignorePatterns <list>',
      'comma-separated list of filename patterns (string literal or regex) to ignore',
      defaultCliCollectOptions.ignorePatterns,
    )
    .option(
      '--no-output',
      'don\'t write to disk but print to console instead',
      defaultCliCollectOptions.output,
    )
    .option(
      '-o, --outDir <path>',
      'outputs generated files to this folder',
      defaultCliCollectOptions.outDir,
    )
    .option(
      '-d, --depth <number>',
      'depth of inDir to collect',
      defaultCliCollectOptions.depth,
    )
    .addOption(
      new Option( '-t, --keyTransform <variant>', 'how to transform route key')
        .choices(['none', 'camelCase'])
        .default(defaultCliCollectOptions.keyTransform)
    )
    .option(
      '-k, --dirkey <name>',
      'how to transform route key',
      defaultCliCollectOptions.dirkey,
    )
    .option(
      '--no-utils',
      'don\'t generate utils for building path with arguments',
      defaultCliCollectOptions.utils,
    )
    .option(
      '--no-typescript',
      'outputs files in javascript instead of typescript',
      defaultCliCollectOptions.typescript,
    )
    .option(
      '--verbose',
      'prints more info during operation',
      defaultCliCollectOptions.verbose,
    )
    .action((rawOptions: CliCollectOptions) => {
      const options: CollectOptions = {
        ...rawOptions,
        extensions: rawOptions.extensions?.split(',') ?? [],
        ignorePatterns: rawOptions.ignorePatterns?.split(',') ?? [],
        depth: Number(rawOptions.depth) || Infinity,
        keyTransform: keyTransformCliToFunc[rawOptions.keyTransform],
      };
      if (options.verbose) {
        console.info('Running "collect" with options:');
        console.table(Object.fromEntries(
          Object.entries(rawOptions).map(
            ([key, value]) => ([
              key,
              { value },
            ])
          )),
        );
      }

      const output = collect(options);
      if (!options.output) {
        console.log(output);
      }
    });
}
