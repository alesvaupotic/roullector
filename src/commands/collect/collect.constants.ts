import type {
  CliCollectOptions,
  CollectOptions,
} from './collect.types';

export function CamelCasify(str: string) {
  return str
    .trim()
    .replace(/([-_\s]+.)/g, function($1) {
      return $1.replace(/[-_\s]/g, '').toUpperCase();
    })
    .replace(/^(.)/, function($1) {
      return $1.toLowerCase();
    });
}

export const defaultCollectOptions: CollectOptions = {
  inDir: 'src/routes',
  extensions: ['.svelte'],
  ignorePatterns: ['^_'],
  outDir: 'src/generated/routing',
  typescript: true,
  keyTransform: CamelCasify,
  verbose: false,
  depth: Infinity,
  dirkey: '__dir',
  utils: true,
};

export const defaultCliCollectOptions: CliCollectOptions = {
  ...defaultCollectOptions,
  extensions: defaultCollectOptions.extensions.join(','),
  ignorePatterns: defaultCollectOptions.ignorePatterns.map((pattern) => (pattern instanceof RegExp) ? pattern.source : pattern).join(','),
  depth: 'Infinity',
  keyTransform: 'camelCase',
};

export const keyTransformCliToFunc: Record<CliCollectOptions['keyTransform'], (str) => string> = {
  camelCase: CamelCasify,
};
