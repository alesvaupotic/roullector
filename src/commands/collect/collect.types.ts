type Modify<T, R> = Omit<T, keyof R> & R;

export type CollectOptions = {
  /** input directory path to collect route data from */
  inDir: string;
  /** file extensions */
  extensions: string[];
  /** patterns to ignore files */
  ignorePatterns: (string|RegExp)[];
  /** route data output directory path */
  outDir: string;
  /** prints more info during operation */
  verbose: boolean;
  /** how to transform route key */
  keyTransform: (key: string) => string;
  /** depth of inDir to collect */
  depth: number;
  /** key to save path for directories with no index file */
  dirkey: string;
  /** generate utils for building path with arguments */
  utils: boolean;
  /** whether to output files in typescript */
  typescript: boolean;
};

export type CliCollectOptions = Modify<CollectOptions, {
  depth: string;
  extensions?: string;
  ignorePatterns?: string;
  keyTransform: 'camelCase';
}>;
