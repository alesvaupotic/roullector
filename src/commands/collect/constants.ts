import type { CollectOptions } from './types';

export const defaultCollectOptions: CollectOptions = {
  pattern: 'src/routes',
  outDir: 'src/generated/routing',
  typescript: true,
};
