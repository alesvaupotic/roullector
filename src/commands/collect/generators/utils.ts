/* eslint-disable no-useless-escape */
import {
  writeFileSync,
  readFileSync,
} from 'fs';
import { resolve } from 'path';

import { transpile } from 'typescript';

import type { CollectOptions } from '$commands/collect/collect.types';

export function generateUtils(options: CollectOptions) {
  const { outDir, typescript } = options;
  let source = '';

  const ts = readFileSync(resolve(__dirname, './route.ts'), 'utf-8');

  if (typescript) {
    source =
`export { default as AppRoutes } from './routes.json';
${ts}
`;
  } else {
    const js = transpile(ts);
    source =
`export { default as AppRoutes } from './routes.json';
}
${js}
`;
  }

  const outputPath = resolve(outDir, 'index.' + (typescript ? 'ts' : 'js'));

  writeFileSync(
    outputPath,
    source,
    { encoding: 'utf-8' },
  );
}
