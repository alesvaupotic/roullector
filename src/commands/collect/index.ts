import { Command } from 'commander';

import { defaultCollectOptions } from './constants';
import type { CollectOptions } from './types';

export function collect() {
  return new Command('collect')
    .description('collect route data')
    .option(
      '-p, --pattern <type>',
      'glob pattern to search for files',
      defaultCollectOptions.pattern
    )
    .option(
      '-o, --outDir <type>',
      'outputs generated files to this folder',
      defaultCollectOptions.outDir
    )
    .option(
      '-t, --typescript',
      'outputs generated files in typescript',
      defaultCollectOptions.typescript
    )
    .action((rawOptions: CollectOptions) => {
      const options = {
        ...defaultCollectOptions,
        ...rawOptions,
      };
    });
}
