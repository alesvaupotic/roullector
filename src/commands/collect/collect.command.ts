import {
  readdirSync,
  mkdirSync,
  existsSync,
  rmSync,
} from 'fs';

import type { CollectOptions } from '$commands/collect/collect.types';
import { generateJSON } from '$commands/collect/generators/json';
import { generateRouteUtil } from '$commands/collect/generators/route';

export function collect(options: CollectOptions) {
  const { outDir } = options;
  try {
    if (!existsSync(outDir)) {
      mkdirSync(outDir, { recursive: true });
    }
    generateJSON(options);
    options.utils && generateRouteUtil(options);
  } catch (error) {
    console.error(error);
    const data = readdirSync(outDir);
    if (data.length === 0) {
      rmSync(outDir, { recursive: true });
    }
  }
}
