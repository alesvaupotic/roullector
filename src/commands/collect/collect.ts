import {
  readdirSync,
  mkdirSync,
  existsSync,
  rmdirSync,
} from 'fs';

import type {
  CollectOptions,
  CollectOutput,
} from '$commands/collect/collect.types';
import { generateJSON } from '$commands/collect/generators/json';
import { generateRouteUtil } from '$commands/collect/generators/route';

import { defaultCollectOptions } from './collect.constants';

export function collect(options: CollectOptions = defaultCollectOptions): CollectOutput {
  const { outDir, output } = options;
  const result = {
    json: null,
    route: null,
  };
  try {
    if (!existsSync(outDir) && output) {
      mkdirSync(outDir, { recursive: true });
    }
    result.json = generateJSON(options);
    if (options.utils) {
      result.route = generateRouteUtil(options);
    }
  } catch (error) {
    console.error(error);
    if (existsSync(outDir) && readdirSync(outDir).length === 0) {
      rmdirSync(outDir, { recursive: true });
    }
    throw error;
  }
  return result;
}
