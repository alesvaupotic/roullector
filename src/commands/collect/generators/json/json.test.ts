import {
  vol,
  fs,
} from 'memfs';

import { defaultCollectOptions } from '$commands/collect/collect.constants';
import type { CollectOptions } from '$commands/collect/collect.types';
import { generateJSON } from '$commands/collect/generators/json';

jest.mock('fs', () => fs);

const defaultOptions: CollectOptions = defaultCollectOptions;

beforeEach(() => {
  vol.fromNestedJSON({
    '__components': {
      'Navbar.svelte': '',
      'Footer.svelte': '',
    },
    '__layout.svelte': '',
    '__error.svelte': '',
    'me.svelte': '',
    'sign-in.svelte': '',
    'sign-up.svelte': '',
    'index.svelte': '',
    'admin': {
      '__components': {
        'Admin.layout.svelte': '',
        'types.ts': '',
      },
      'users': {
        '[id]': {
          'index.svelte': '',
          'types': '',
        },
        'assets': {
          '__layout.svelte': '',
          'index.svelte': '',
          'photos': {
            '[id]': {
              'types': '',
              'index.svelte': '',
            },
          },
          'posts': {
            'youtube': {
              'types': '',
              '[id].svelte': '',
              '[slug].svelte': '',
            },
          },
        },
      },
    }
  }, 'src/routes');
});

afterEach(() => {
  vol.reset();
});

const generateJSONWithOptions = (snapshotToMatch: string, options: Partial<CollectOptions> = {}) => {
  const outputPath = generateJSON({
    ...defaultOptions,
    ...options,
  });
  const generated = fs.readFileSync(outputPath, 'utf-8');
  expect(generated).toMatchSnapshot(snapshotToMatch);
};

test(`generate JSON but empty directory`, () => {
  vol.reset();
  const outputPath = generateJSON(defaultOptions);
  expect(outputPath).toBe(null);
});

test(`generate JSON with default options`, () => {
  generateJSONWithOptions('generateJSON: default');
});

test(`generate JSON with a custom extension matcher`, () => {
  generateJSONWithOptions('generateJSON: custom extension matcher', { extensions: ['.ts'] });
});

test(`generate JSON with depth of 1`, () => {
  generateJSONWithOptions('generateJSON: depth 1', { depth: 1 });
});

const ignorePatterns = ['^_', '^assets'];
test(`generate JSON with custom ignorePatterns: ${ignorePatterns}`, () => {
  generateJSONWithOptions('generateJSON: custom ignorePatterns',{ ignorePatterns });
});

const dirkey = '__customDir__';
test(`generate JSON with custom dirkey: ${dirkey}`, () => {
  generateJSONWithOptions('generateJSON: custom dirkey', { dirkey });
});
