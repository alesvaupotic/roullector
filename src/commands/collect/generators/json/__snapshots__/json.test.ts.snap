// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate JSON with a custom extension matcher: generateJSON: custom extension matcher 1`] = `"{}"`;

exports[`generate JSON with custom dirkey: __customDir__: generateJSON: custom dirkey 1`] = `
"{
  \\"admin\\": {
    \\"users\\": {
      \\"[id]\\": {
        \\"index\\": \\"/admin/users/[id]\\"
      },
      \\"assets\\": {
        \\"index\\": \\"/admin/users/assets\\",
        \\"photos\\": {
          \\"[id]\\": {
            \\"index\\": \\"/admin/users/assets/photos/[id]\\"
          },
          \\"__customDir__\\": \\"/admin/users/assets/photos\\"
        },
        \\"posts\\": {
          \\"youtube\\": {
            \\"[id]\\": \\"/admin/users/assets/posts/youtube/[id]\\",
            \\"[slug]\\": \\"/admin/users/assets/posts/youtube/[slug]\\",
            \\"__customDir__\\": \\"/admin/users/assets/posts/youtube\\"
          },
          \\"__customDir__\\": \\"/admin/users/assets/posts\\"
        }
      },
      \\"__customDir__\\": \\"/admin/users\\"
    },
    \\"__customDir__\\": \\"/admin\\"
  },
  \\"index\\": \\"/\\",
  \\"me\\": \\"/me\\",
  \\"signIn\\": \\"/signIn\\",
  \\"signUp\\": \\"/signUp\\"
}"
`;

exports[`generate JSON with custom ignorePatterns: ^_,^assets: generateJSON: custom ignorePatterns 1`] = `
"{
  \\"admin\\": {
    \\"users\\": {
      \\"[id]\\": {
        \\"index\\": \\"/admin/users/[id]\\"
      },
      \\"__dir\\": \\"/admin/users\\"
    },
    \\"__dir\\": \\"/admin\\"
  },
  \\"index\\": \\"/\\",
  \\"me\\": \\"/me\\",
  \\"signIn\\": \\"/signIn\\",
  \\"signUp\\": \\"/signUp\\"
}"
`;

exports[`generate JSON with default options: generateJSON: default 1`] = `
"{
  \\"admin\\": {
    \\"users\\": {
      \\"[id]\\": {
        \\"index\\": \\"/admin/users/[id]\\"
      },
      \\"assets\\": {
        \\"index\\": \\"/admin/users/assets\\",
        \\"photos\\": {
          \\"[id]\\": {
            \\"index\\": \\"/admin/users/assets/photos/[id]\\"
          },
          \\"__dir\\": \\"/admin/users/assets/photos\\"
        },
        \\"posts\\": {
          \\"youtube\\": {
            \\"[id]\\": \\"/admin/users/assets/posts/youtube/[id]\\",
            \\"[slug]\\": \\"/admin/users/assets/posts/youtube/[slug]\\",
            \\"__dir\\": \\"/admin/users/assets/posts/youtube\\"
          },
          \\"__dir\\": \\"/admin/users/assets/posts\\"
        }
      },
      \\"__dir\\": \\"/admin/users\\"
    },
    \\"__dir\\": \\"/admin\\"
  },
  \\"index\\": \\"/\\",
  \\"me\\": \\"/me\\",
  \\"signIn\\": \\"/signIn\\",
  \\"signUp\\": \\"/signUp\\"
}"
`;

exports[`generate JSON with depth of 1: generateJSON: depth 1 1`] = `
"{
  \\"index\\": \\"/\\",
  \\"me\\": \\"/me\\",
  \\"signIn\\": \\"/signIn\\",
  \\"signUp\\": \\"/signUp\\"
}"
`;
